### ISSUE TITLE
gstreamer encoder

### ISSUE THREAD

**Issue by totaam** [Xpra-org/xpra#3706]
Worth a try.
We already have `appsink` and `appsrc` working code so it should be possible to feed `BGRX` or `YUV420P` to a video encoder (initially `x264`) and get the stream as output.
This can be used for testing hardware encoders and see what the difference is with `libva` (#451): perhaps this one won't lock up the system quite so easily?

If that pans out, we can try to feed the pixel data to the pipeline without copying it.
It should be possible to implement this as an element in python:
[Implementing an audio plotter](https://mathieuduponchelle.github.io/2018-02-15-Python-Elements-2.html)
[How to use Gstreamer AppSrc in Python](https://lifestyletransfer.com/how-to-use-gstreamer-appsrc-in-python/)

As for the pipeline option, there are some examples of gstreamer syntax that match what we already do with `x264` in Cython code:
[Realtime/zero-latency video stream: what codec parameters to use?](https://stackoverflow.com/questions/30730082/)

Element references:
* [x264](https://gstreamer.freedesktop.org/documentation/x264/index.html?gi-language=c)
* [appsink](https://gstreamer.freedesktop.org/documentation/app/appsink.html#appsink)
* [appsrc](https://gstreamer.freedesktop.org/documentation/app/appsrc.html?gi-language=c#appsrc)

Some information on EOS handling: [How to wait for x264enc to encode buffered frames on end-of-stream](https://stackoverflow.com/questions/57575951/)

**Comment by totaam** [comment]
Works pretty well!
Still TODO:
* seems to be re-cycled too often - and this is more expensive than with the plain `x264` encoder
* define gpu / cpu costs in encoder spec
* find image input size limits dynamically
* avoid concatenating the YUV planes into a single buffer somehow? (or perhaps  hack it so that `csc_libyuv` spits out `YUV420` without the `P` in a single buffer, unpadded)
* verify that [Gst.Buffer.new_wrapped_full](https://lazka.github.io/pgi-docs/Gst-1.0/classes/Buffer.html#Gst.Buffer.new_wrapped_full) does what we want, without leaking - referencing memory without refcounting it
* test on more hardware - even on AMD?
* test with other elements: `nvh264enc` / `nvautogpuh264enc` / `nvcudah264enc`, 
* this element only accepts `I420` input - how can we use even pseudo-lossless without `444`?
* tune encoder settings based on our `speed` and `quality` knobs
* try to support `B`-frames?
* implement simple decoder? ie: [nvh264dec](https://gstreamer.freedesktop.org/documentation/nvcodec/nvh264dec.html?gi-language=c)? could support hardware decoding quite easily - but not sure about opengl buffer sharing..
* implement a csc module using [videoconvert](https://gstreamer.freedesktop.org/documentation/videoconvertscale/videoconvert.html?gi-language=c)? or even [cudaconvert](https://gstreamer.freedesktop.org/documentation/nvcodec/cudaconvert.html?gi-language=c)

**Comment by totaam** [comment]
The commits above allows us to add `av1` encoding almost for free, and `h264` via avcodec.
But there are also some new issues:
* the performance of these encoders is vastly different: `av1enc` is unbearably slow..
* out private ffmpeg library conflicts with the one gstreamer links against... perhaps time to drop it? (RPM distros still have `libyuv` for csc and the new `gstreamer` codecs are already more useful than the `ffmpeg` encoder - the only thing that would then be missing is the `h264` `avcodec` decoder - perhaps the gstreamer avcodec decoder can take its place if we find a way to avoid copying the `YUV` pixel data?)
* not all encoders can be tuned for `zerolatency` and not buffer frames..

**Comment by totaam** [comment]
This is likely to sway the decision of how to deal with `ffmpeg` conflicts between our private library and the one used by `gstreamer`: F37 Xorg server processes crashes with nvidia drivers installed and the gstreamer encoder:

```
xpra start --no-daemon :200
(..)
2023-01-06 10:32:12,756 31.3GB of system memory
Xorg: ../../../include/privates.h:121: dixGetPrivateAddr: Assertion `key->initialized' failed.
(EE) 
(EE) Backtrace:
(EE) 0: /usr/libexec/Xorg (OsLookupColor+0x13d) [0x5649171fe0bd]
(EE) 1: /lib64/libc.so.6 (__sigaction+0x50) [0x7f35b5aa2b50]
(EE) 2: /lib64/libc.so.6 (__pthread_kill_implementation+0x10c) [0x7f35b5af2e7c]
(EE) 3: /lib64/libc.so.6 (gsignal+0x16) [0x7f35b5aa2aa6]
(EE) 4: /lib64/libc.so.6 (abort+0xd7) [0x7f35b5a8c7fc]
(EE) 5: /lib64/libc.so.6 (__assert_fail_base.cold+0xf) [0x7f35b5a8c71b]
(EE) 6: /lib64/libc.so.6 (__assert_fail+0x46) [0x7f35b5a9b656]
(EE) 7: /usr/libexec/Xorg (SrvXkbResizeKeyType+0x825) [0x5649171c6a25]
(EE) 8: /usr/libexec/Xorg (DRI2Authenticate+0xb5) [0x5649171c70d5]
(EE) 9: /usr/libexec/Xorg (DRI2ScreenInit+0x8e1) [0x5649171c9b91]
(EE) 10: /usr/libexec/Xorg (miPutImage+0x163d) [0x56491708bdb4]
(EE) 11: /lib64/libc.so.6 (__libc_start_call_main+0x80) [0x7f35b5a8d510]
(EE) 12: /lib64/libc.so.6 (__libc_start_main+0x89) [0x7f35b5a8d5c9]
(EE) 13: /usr/libexec/Xorg (_start+0x25) [0x56491708c0f5]
(EE) 
(EE) 
Fatal server error:
(EE) Caught signal 6 (Aborted). Server aborting
(EE) 
(EE) 
Please consult the The X.Org Foundation support 
	 at http://wiki.x.org
 for help. 
(EE) Please also check the log file at "./Xorg.log" for additional information.
(EE) 
(EE) Server terminated with error (1). Closing log file.
X connection to :200 broken (explicit kill or server shutdown).
X connection to :200 broken (explicit kill or server shutdown).
```

**Comment by totaam** [comment]
Interesting, it may soon be possible to use `vaapi` via WSL2: [Microsoft Officially Launches D3D12 GPU Video Acceleration For WSL Linux Use](https://www.phoronix.com/news/WSL-D3D12-GPU-Video-Accel)

**Comment by totaam** [comment]
New issue: https://github.com/Xpra-org/xpra/issues/3744#issuecomment-1441575207
`video_orc_unpack_I420` `Cannot access memory at address` could mean that the buffer we supply is too small.
Perhaps the output from `csc_libyuv` doesn't have the exact padding / rounding expected by gstreamer - I thought it did..

**Comment by TijZwa** [comment]
Full backtrace for #3744 
[gdb.txt](https://github.com/Xpra-org/xpra/files/10930640/gdb.txt)

**Comment by totaam** [comment]
@TijZwa is this easily reproducible?
If so, can you post the server log running with `-d video,gstreamer`?

**Comment by totaam** [comment]
Do we need to set the [GstVideoColorimetry](https://gstreamer.freedesktop.org/documentation/video/video-color.html?gi-language=c#GstVideoColorimetry) to full range? [ffmpeg does it](https://github.com/FFmpeg/FFmpeg/commit/7ca71b79f2b3256a0eef1a099b857ac9e4017e36).
`gst_video_color_range_offsets`: https://github.com/GStreamer/gst-plugins-base/blob/ce937bcb21412d7b3539a2da0509cc96260562f8/gst-libs/gst/video/video-color.c#L204
Set [GstVideoInfo](https://gstreamer.freedesktop.org/documentation/video/video-info.html?gi-language=c) using caps?

For [x264enc](https://gstreamer.freedesktop.org/documentation/x264/index.html?gi-language=c), we should also handle bandwidth limits by switching to [cbr mode] (https://gstreamer.freedesktop.org/documentation/x264/index.html?gi-language=c#GstX264EncPass) and still set [qp-max](https://gstreamer.freedesktop.org/documentation/x264/index.html?gi-language=c#x264enc:qp-max)?
Only use `intra-refresh` when bandwidth is limited? (with lossy transports?)

**Comment by totaam** [comment]
We can also do `nvenc` via `gstreamer`, but this is currently disabled by default:
https://github.com/Xpra-org/xpra/blob/4d62441f35b0014ab61157ca4ae5f02f9cc4d4d4/xpra/codecs/gstreamer/encoder.py#L30
Because when we hit the artificial context limit on consumer cards, the gstreamer pipeline will just timeout instead of erroring out quickly - which is problematic. The error can be seen at higher debug logging levels.

**Comment by algeorge** [comment]
I'd like to check the current status of Xpra encoding and decoding with Intel QuickSync. I could not find any information on the internet as this is a rapidly moving project.

Does it work via vaapi through gstreamer?

**Comment by totaam** [comment]
@algeorge the gstreamer encoder in this ticket is included in all v5 official builds (currently in the beta area only) but it is disabled by default:
https://github.com/Xpra-org/xpra/blob/0caa26f82d80de7e5317e75b4e1604db6504dc10/fs/etc/xpra/conf.d/30_picture.conf.in#L26
In the future, I would like to add utilities to make it easier for end users to test and enable the pipeline combinations that work best on their specific hardware.
For now, this is a manual process.

**Comment by algeorge** [comment]
@totaam

Manual is great! Because it even works!

As far as settings to pass through, are there any options for gstreamer anywhere? To modify colorspace, bit-depth, quality, bitrate... etc. Or is that handled as part of the regular Xpra commands.

**Comment by algeorge** [comment]
> As for the pipeline option, there are some examples of gstreamer syntax that match what we already do with x264 in Cython code:
[Realtime/zero-latency video stream: what codec parameters to use?](https://stackoverflow.com/questions/30730082/)

I would like to add that this depends on the encoder which is why I am so interested in QuickSync. MJPEG has a sub 1ms latency for encoding with a fairly high bitrate. MPEG2 is in the same ballpark with latency because the codec is so simple.

**Comment by totaam** [comment]
We used to have MPEG1 and MPEG2 support via the ffmpeg encoder, but this was removed recently.
The trade-offs and limitations were just no longer justifiable compared to modern codecs.
For tuning the ffmpeg encoder, we can use the link you posted and also this one: https://github.com/selkies-project/selkies-gstreamer/blob/master/src/selkies_gstreamer/gstwebrtc_app.py as a base.
But the key thing that allows xpra to perform so well is its ability to dynamically tune encoder settings to better adapt to content and bandwidth conditions.
This is more tricky to implement when the exact pipelines vary. This will take time.
